# 함수

코딩의 기본은 **중복 제거**이다. 
중복을 가능한 한 피하면서 구현한 코드는 아래의 성질을 만족한다.

* <font color='blue'>코드의 양이 줄어든다.</font>
* <font color='blue'>수정하기 편하다.</font>
* <font color='blue'>이해하기 쉽다.</font>
* <font color='blue'>재활용이 용이하다.</font>

실제로 모든 프로그래밍언어는 일반 사용자가 위 조건을 충족시키는 프로그램을 구현하도록
도와주는 많은 장치와 도구를 함께 제공한다. 
지금까지 살펴 본 변수, 조건문, 다양한 자료구조(문자열, 불리언, 컨테이너 등등), 반복문 모두
보다 효율적으로 버그가 없는 프로그램을 구현하도록 도와주는 장치와 도구들이다.
이외에 함수, 모듈, 클래스와 객체 등 다른 장치와 도구들이 있으며 앞으로 하나씩 살펴볼 것이다.

프로그래밍을 배우는 사용자의 입장에서는 그런 장치와 도구들을 어떻게 잘 사용할 수 있는지를 
경험하고 연습하는 것이 무엇보다도 중요하다.

이번 장에서는 함수에 대해 간략하게 다룬다.

### 참고 사항
아래 내용은 
[Think Python 3장](http://greenteapress.com/thinkpython2/html/thinkpython2004.html) 
내용을 번역 및 요약수정한 내용이다.

파이썬을 이용한 설명이지만 루비, 자바 등 어떤 언어에 대해서도 동일한 내용의 예제를 이용한
설명이 가능하다.

## 값과 함수

프로그램을 구현한다는 것은 기본적으로 특정 종류의 값들을 조작하여 원하는 값을 구하는 과정을
특정 프로그래밍언어로 표현하는 것이다. 

지금가지 여러 종류의 값들을 살펴 보았다.
* 정수: -1, 0, 2, -3, ...
* 실수: 0.25, -3.54, ...
* 불리언 값: True, False
* 문자열: "python", "a3ge", ...
* 컨테이너 등등

모든 프로그래밍언어마다 값을 표현하는 고유의 방식이 있다.
정수와 실수는 거의 모든 언어에서 표현하는 방식이 동일하지만 이외의 자료형에 속하는 값들은 비슷하지만 조금씩 다르게
구현되었으며, 따라서 사용되는 방식도 그에 따라 조금씩 다를 수 있다.

따라서 특정 자료형을 어떻게 표현하는지 정확하게 알고 있어야 하며,
문법에 맞지 않게 사용할 경우 기본적으로 오류가 발생한다. 
여기서는 함수를 이용하여 값을 표현하는 방식을 알아본다. 

먼저 함수가 무엇인지를 알아본다.

## 함수 이해하기

함수가 하는 일은 매우 간단하다. 
특정 값들을 먼저 받아들인 다음에 그 값을 조작하여 어떤 값을 되돌려준다. 

아래 그림에서 절대값을 되돌려주는 `abs` 함수와 두 숫자의 합을 되돌려주는 `add` 함수를 예를 들어 
함수를 설명하고 있다.
기본적인 아이디어는 왼쪽에 입구를 통해 값을 받아들이고, 오른쪽 출구를 통해 값을 전달하는 것이다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_name.png" style="width:300px">
</td>
</tr>
</table>
</p>

* `abs` 함수: 인자로 사용된 숫자의 절대값을 되돌려주는 함수

In [1]:
abs(-3)

3

* `add` 함수: 일상적으로 덧셈(`+`) 기호로 표시되는 함수의 원래 이름이다.
    `add` 함수를 사용하려면 `operator` 모듈을 임포트해야 한다.

In [2]:
from operator import add

In [3]:
add(3, 5) == 3 + 5

True

### 매개변수와 인자
위 그림에서 `number`, `left`, `right` 등은 함수의 인자를 받아들이는데 사용되는 **매개변수**들을 나타낸다. 
매개변수를 통해 함수에게 전달되는 값들을 **인자**라고 부른다. 
인자의 개수는 매개변수의 개수와 일치해야 한다.

함수와 매개변수들의 이름은 각각의 역할에 맞게 정하는 것을 권유한다.
그러면 함수와 매개변수들의 이름을 보고 함수와 각 매개변수들의 의미와 역할을 파악하는 데에 보다 유리하다.

### 지금까지 사용한 함수

앞서 살펴 본 수학함수 이외에 많은 함수들이 있으며 
이미 여러 함수들을 살펴보았다.
`type`, `int`, `float`, `str`, `input`, `time.sleep`, `urllib.request.urlopen`, 등등

In [4]:
type(42)

int

In [5]:
int('32')

32

In [6]:
float('3.9999')

3.9999

In [7]:
str(32)

'32'

In [8]:
type(str(32))

str

### 함수와 추상화
도구가 이미 준비되어 있을 때 그 도구가 어떻게 만들어졌는지 알 필요 없듯이
이미 정의된 함수가 어떻게 정의되어 있는가 또한 굳이 알 필요 없다.
이와 같은 함수의 특성을 추상성이라 하며, 함수를 활용한 추상화를 적절하게 활용하는 것이
프로그래밍의 핵심에 포함된다.

### 도구로서의 함수
도구의 사용법을 알아야 하는 것처럼 함수의 경우는 
함수를 언제 어떻게 사용하는 가를 반드시 알아야 한다.
일반적으로 함수에 대해서는 이정도만 알고 있어도 충분하다.

### 새 함수 정의하기

앞서 언급한 함수들은 파이썬에서 이미 누군가에 의해 정의된 함수들이다.
이런 함수들을 파이썬 **내장함수**라 부른다.

하지만 일반 사용자가 필요에 따라 함수를 임의로 정의할 수 있다.
새 함수를 정의하는 양식은 다음과 같다.
```python
def 함수이름(인자1, 인자2, ...):
    코드
```

#### 예제

아래 시간을 분으로 변환해서 되돌려주는 함수를 아래와 같이 정의할 수 있다.

In [9]:
def hour_to_min(hour):
    minutes = hour * 60
    return minutes

`hour_to_min` 함수는 `hour`라는 매개변수 하나만을 사용하며 인자를 받으면 분 단위로 변경한 값을
담고 있는 `minutes` 변수의 값을 되돌려 준다.

함수가 되돌려 준 값은 다른 함수의 인자로 사용되거나 변수를 선언하는 값으로 활용될 수 있다. 
이와 같이 함수가 되돌려 주는 값을 필요에 따라 재활용할 수 있다.

예를 들어, 2시간이 몇 분인지 알고자 하면 `hour_to_min` 함수를 아래와 같이 사용하면 된다.

In [10]:
two_hour = hour_to_min(2)
print("두 시간은", two_hour, "분에 해당합니다.")

두 시간은 120 분에 해당합니다.


위 코드에서 `hour_to_min(2)`을 실행하여 되돌려 받은 값은 `two_hour` 변수에 할당하였고
그 변수를 `print` 함수의 둘째 인자로 활용하였다.

### 지역변수와 전역변수
함수를 선언할 때 사용되는 매개변수와 함수 본체에서 선언되는 변수는 함수가 실행되는 동안에만 의미를 갖는 변수들이며,
이런 변수들을 **지역변수**라 부른다. 
지역변수 가인 변수들은 **전역변수**라 부른다.

예를 들어, `hour_to_min` 함수를 정의할 때 사용된 `hour`와 본체에서 선언된 `minutes` 변수는 모두 지역함수이며,
`two_hour` 는 함수 밖에서 선언된 전역변수이다.

지역변수들은 함수 밖에서는 어떤 의미도 갖지 않는다. 예를 들어, 아래 코드를 실행하면 오류가 발생한다.

In [11]:
two_hour = hour_to_min(2)
print(minutes)

NameError: name 'minutes' is not defined

물론 아래의 경우도 오류가 발생한다.

In [12]:
two_hour = hour_to_min(2)
print(hour)

NameError: name 'hour' is not defined

위에서 오류가 발생하는 이유는 `hour_to_min` 함수가 인자 2와 함께 실행되어 종료가 되면 
실행도중에 선언되어 사용된 `hour`와 `minutes` 변수의 의미도 완전히 사라지기 때문이다.

아래 **함수호출**에 대한 설명에서 지역변수가 프로그램 실행 도중에 어떻게 의미를 갖거나 잃어버리는가를
**프레임** 개념을 이용하여 좀 더 자세히 설명될 것이다.

### 함수의 리턴값

`hour_to_min` 와 같은 함수가 되돌려주는 값을 일반적으로 **리턴(return)값**이라 부른다.
리턴 값은 함수 정의 본체에서 `return`이란 키워드 바로 뒤에 오는 값을 의미하며, 
함수가 특정 인자와 함께 실행되면 함수 본체의 코드에 맞추어 값을 조작한 후에 
최종적으로 되돌려주는 값을 가리킨다.

파이썬의 경우 리턴값이 앞서 설명한 대로 `return` 키워드와 함께 명시된 함수와 
그렇지 않은 함수가 존재한다. 

#### 리턴값이 명시된 함수
* `hour_to_min` 함수의 경우처럼 `return` 키워드가 명시적으로 사용된 함수.
* 함수를 실행하다가 `return`을 만나게 되면 리턴값을 되돌려 준 후 함수의 실행을 바로 멈춘다.

C 또는 Java 언어와는 달리 파이썬에서는 함수의 리턴값의 자료형을 명시하지 않는다.

#### 예제
`hour_to_min` 함수를 C언어로 구현한다면 다음과 같은 모양을 갖는다.
```c
int hour_to_min(int hour) {
    int minutes = hour * 60;
    return minutes;
}
```

#### 예제
`hour_to_min` 함수를 Java 언어로 구현한다면 다음과 같은 모양을 갖는다.
```java
public int hour_to_min(int hour) {
    int minutes = hour * 60;
    return minutes;
}
```

#### 리턴값이 명시되지 않은 함수
대표적으로 `print` 함수가 리턴값이 명시되지 않은 함수이다. 
리턴값이 명시되지 않은 함수의 리턴값은 항상 `None` 이다.
`None`은 사실상 아무런 의미를 갖지 않는 값이며 어떠한 재활용도 불가능하다.

예를 들어, 아래 코드는 `print` 함수가 리턴하는 값을 확인하고자 하지만 말했듯이 `None`이란 것만 확인할 수 있다.

In [13]:
x = print(1)
print(x)

1
None


**주의:** 위 코드를 실행한 결과의 첫째 줄에서 보이는 `1`은 `print` 함수의 리턴값이 아니라,
`print(1)`을 실행하여 모니터에 숫자 1을 출력하는 `print` **함수의 부수효과**에 불과하다.
함수의 부수효과에 대해서는 아래에서 좀 더 자세히 다룬다.

#### 예제
`hour_to_min` 함수를 아래와 같이 리턴값이 명시되지 않은 함수로 정의할 수도 있다.

In [14]:
def hour_to_min_noReturn(hour):
    minutes = hour * 60
    print(minutes)

위 함수를 이용하여 두 시간을 분단위로 확인하고자 하면 다음과 같다.

In [15]:
hour_to_min_noReturn(2)

120


위 코드를 보면 마치 120을 리턴하는 것처럼 보이지만 사실은 `print` 함수가 모니터에 출력한 결과를 
우리에게 보여주는 거에 불과하다.
실제로 아래와 같이 리턴값을 활용하려 하면 제대로 작동하지 않는다.

In [16]:
two_hour_noReturn = hour_to_min_noReturn(2)
print(two_hour_noReturn)

120
None


실제로 `hour_to_min_noReturn(2)`의 자료형은 `None`의 자료형인 `NoneType`이다.

In [17]:
type(hour_to_min_noReturn(2))

120


NoneType

#### 예제
`hour_to_min` 함수를 C언어로 구현한다면 다음과 같은 모양을 갖는다.
```c
void hour_to_min(int hour) {
    int minutes = hour * 60;
    printf(minutes);
}
```

#### 예제
`hour_to_min_noReturn` 함수를 Java 언어로 구현한다면 다음과 같은 모양을 갖는다.
```java
public void hour_to_min(int hour) {
    int minutes = hour * 60;
    System.out.println(minutes);
}
```

#### 함수의 리턴값은 단 하나

함수 본체 코드에 `return` 문이 여러 번 사용되었다 하더라도 함수는 하나의 값만 리턴한다.

예를 들어, 아래 루비에서 정의된 `login` 함수 본체에는 `return` 문이 두 번 사용되었다.

```ruby
def login(_id)
  members = ['egoing', 'k8805', 'leezche']
  for member in members do
      if member == _id
          return true
      end
  end
  return false
end
```

하지만 `login` 함수가 호출되어 실행되는 과정에 두 리턴문 문 중에 하나의 리턴문이 실행되면
바로 지정된 리턴값을 반환하면서 함수의 실행이 멈추게 되어 다른 리턴문은 실행되지 않게 된다.

## 함수 표현식

함수를 이용해서 값을 표현하는 방식 **함수 표현식**이라 부른다. 

예를 들어, 제곱근을 생성하는 함수인 `sqrt`의 경우 실수 하나를 인자로 받아서 그 실수의 제곱근을 생성하여 돌려준다.
`sqrt` 함수를 이용하여 실수 `r`의 제곱근을 아래와 같이 표현할 수 있다.
```python
sqrt(r)
```    
또한, 실수 `r`의 제곱근에다가 `2`를 더한 값은 다음과 같이 표현된다.
```python
add(sqrt(r),2)
```    

#### 주의사항
앞서 언급한 대로 `sqrt`를 이용한 함수 표현식이 `add` 함수의 첫째 인자로 활용되었다.
이런 표현식은 **함수의 합성**을 활용한 표현식이다.
이 경우 `add` 함수의 첫째 인자값으로 `sqrt(r)`이 실행되어 되돌려진 값, 즉, `r`의 제곱근이 사용된다.

## 함수호출

함수 표현식은 특정 값을 표현하기 위해 사용된다.
하지만 그 표현식이 어떤 값을 나타내는지를 알려면 사용된 함수들을 실행해야만 하며,
이렇게 함수 표현식을 실행하는 것을 **함수호출**이라 부른다.

예를 들어, `sqrt(8)`은 `8`의 제곱근을 나타내는 값이지만 실제로 `8`의 제곱근이 정확히 얼마인지는 말하지 않는다.
`8`의 제곱근이 얼마인지를 확인하기 위해서는 이제 `sqrt`에 `8`을 인자로 입력하여 호출해야 한다.

"`sqrt`에 `8`을 인자로 입력하여 호출한다" 란 말이 좀 복잡하게 들리지만 실제로는 매우 간단하다.
바로 아래 명령을 실행하기만 하면 된다.
```python
sqrt(8)
```

#### 주의사항
    
`sqrt`를 사용하려면 먼저 `math` 모듈을 임포트해야 한다. 
파이썬의 특정 모듈에서 특정 함수를 임포트 하는 방식은 다음과 같이 할 수 있다.
```python    
from math import sqrt
```

In [18]:
from math import sqrt

sqrt(8)

2.8284271247461903

즉, `sqrt(8)`의 정확한 값은 `2.8284271247461903`이다.
8의 제곱근이 무리수이기 때문에 근사값에 불과하다.
하지만 8의 제곱근의 정확한 값에 매우 근사한 값이다.

이와 같은 방식으로 `add(sqrt(r),2)`의 정확한 값을 계산할 수 있다.

In [19]:
add(sqrt(8), 2)

4.82842712474619

이와 같이 함수호출 표현식의 정확한 값을 계산하는 과정을 **함수호출**이라 부른다.
즉, 함수를 정말로 실행하여 특정 값을 얻어내는 과정이 함수호출이다.

### 함수호출 실행과정

함수호출 표현식을 실행하여 결과값을 확인하는 것을 함수호출이라 하였다.
여기서는 함수호출이 실행되는 과정을 예제들을 통해 좀 더 자세히 살펴본다. 

#### 예제

아래 함수호출 표현식의 결과값이 어떤 순서대로 정해지는가를 확인해보자.

**주의:** 
* `add` 함수와 마찬가지로 `mul` 함수는 일상적으로 곱셈(`*`) 기호로 표시되는 함수의 원래 이름이며,
    `operator` 모듈에 포함되어 있다.
* `0x`로 시작하는 숫자는 16진법을 의미한다.
* `0o`로 시작하는 숫자는 8진법을 의미한다.

    

In [20]:
from operator import mul

In [21]:
mul(add(2, mul(0x4, 0x6)), add(0x3, 0o5))

208

* `mul` 함수의 정의를 확인한다.
* `mul`의 첫 번째 인자인 `add(2, mul(0x4, 0x6))`를 확인하고 함수호출을 실행한다.
    + `add` 함수의 정의를 확인한다. 
    + `add`의 첫 번째 인자인 `2`를 확인한다. `2`는 더 이상 계산할 필요가 없다.
    + `add`의 두 번째 인자닌 `mul(0x4, 0x6)`를 확인하고 함수호출을 실행한다. 
        - `mul` 함수의 정의를 확인한다. 
        - `mul`의 첫 번째 인자인 `0x4`를 확인한다. `0x4`는 16진법으로 표시된 정수 4임을 확인한다.
        - `mul`의 두 번재 인자인 `0x6`를 확인한다. `0x6`는 16진법으로 표시된 정수 6임을 확인한다.
        - `mul(4,6)`을 실행하여 `24`를 리턴한다.
    + `add(2, 24)`를 실행하여 `26`을 리턴한다.
* `mul`의 두 번째 인자인 `add(0x3, 0o5)`를 확인한다. 
    + `add` 함수의 정의를 확인한다.
    + `add`의 첫 번째 인자인 `0x3`를 확인한다. `0x3`는 16진법으로 표시된 정수 3임을 확인한다.
    + `add`의 두 번째 인자인 `0o5`를 확인한다. `0o5`는 8진법으로 표시된 정수 5임을 확인한다.
    + `add(3, 5)`실행하여 `8`을 리턴한다.
* `mul(26, 8)`을 실행하여 `208`을 리턴한다.

위 설명에서 들여쓰기에 주의해야 한다.
더 들여서 쓴 문장이 모두 실행이 된 뒤에야 그 아랫부분에서 덜 들여서 쓴 문장이 실행된다. 
그림으로 나타내면 다음과 같다.

아래 그림에 네모상자로 둘러싸인 부분이 현재 실행중인 함수호출 또는 함수호출의 결과값을 나타낸다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_call.png" style="width:500px">
</td>
</tr>
</table>
</p>

## 부수효과가 없는 함수와 있는 함수

프로그램에서 사용되는 함수를 구분하는 다양한 기준이 있다.
여기서는 부수기능의 존재여부에 따른 함수 분류 기준을 살펴본다.

### 부수효과가 없는 함수 (Pure functions)

함수가 호출되어 실행되어 실행결과를 돌려줄 때까지 실행결과를 생성하기 위한 일 이외에 다른 어떤 일도 하지 않는다면 그 함수를 부수효과가없는 함수라 부른다.

예를 들어 절대값을 계산하는 `abs` 함수와 덧셈을 행하는 `add` 함수 등이 순수한 함수이다. 
아래 그림에서 보듯이 인자를 입력받은 후에 각각 절대값과 덧셈을 실행하여 결과값을 돌려주는 일 이외에는 다른 일을 행하지 않는다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_pure.png" style="width:250px">
</td>
</tr>
</table>
</p>

### 부수효과가 있는 함수 (Impure functions)

부수효과가 있는 함수는 부수효과가 없는 함수와는 달리 결과값을 돌려주는 것 이외에 부수적인 일을 한다.
대표적으로 파이썬 내장함수인 `print` 함수가 부수효과를 갖는 함수이다.

그런데 `print`가 부수효과를 갖는 함수라는 것을 확인하려면 아래 두 가지를 알아야 한다.

* 리턴하는 값이 무엇인가?
* 부수적으로 어떤 일을 하는가?

먼저, `print`가 리턴하는 값은 널 값(null value)이다. 
널 값은 말 그대로 어무런 내용도 없는 값을 의미한다.
즉, 값은 값인데 아무 데도 사용할 수 없는 값을 의미한다.
즉, 무늬만 값인 것이다. 
파이썬은 이러한 널값을 `None`으로 표시한다.

다음으로, `print` 함수가 부수적으로 하는 일은 예를 들어 터미널 창에 어떤 문자열을 출력하는 것이다.
`print("Hello Python")` 방식으로 `print` 함수를 호출하면 아래와 같이 `'Hello Python'` 이란 문자열을 출력한다.

In [22]:
print("Hello Python")

Hello Python


`print("Hello Python")`을 실행하였을 때 보여주는 `'Hello Python'`은 실행 결과값이 아니라 `print` 함수가 부수적으로 하는 일이라는 점을 절대적으로 명심해야 한다.

#### 예제: 부수효과를 갖는 함수의 호출과정

부수효과를 갖는 함수호출의 실행과정은 부수효과가 없는 함수의 호출과정과 기본적으로 동일하다.
하지만 중간중간에 부수적인 일도 함께 벌어진다.

먼저 아래 코드의 실행결과가 어떻게 도출되었는가를 잘 생각해보자.

In [23]:
print(print(1), print(2))

1
2
None None


아래의 그림은 위 코드를 호출하는 과정을 설명한다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/fun_print.png" style="width:350px">
</td>
</tr>
</table>
</p>

**설명:** 기본적으로 앞서 설명한 방식으로 진행되지만 3번 사용된 각각의 `print` 함수 호출과정에서 
특정값을 부수적으로 화면에 출력한다. 

* 가장 먼저 첫 번째 인자에 사용된 `print` 함수에 의해 `1`을 출력하고
* 다음으로 두 번째 인자에 사용된 `print` 함수에 의해 `2`를 출력하고
* 마지막으로 제일 먼저 호출된 `print` 함수에 의해 각 인자의 리턴값으로 이루어진 `None None`을 출력한다. 

## 함수호출 실행과정과 프레임 이해

함수를 호출하여 실행을 하면 컴퓨터 메모리의 스택(stack) 영역에 함수의 호출과 실행과정, 그리고 실행결과값과 관련된
정보의 변화가 일어난다. 
이런 정보의 변화를 **프레임(frame)** 이라 부르며, 프레임은 모든 함수와 변수의 생성, 호출, 삭제 과정을 기록한다.

또한 하나의 함수가 호출될 때마다 하나의 프레임이 임시로 생성되었다가 함수의 실행이 종료되면 해당 함수의
리턴값을 다른 프레임에 넘겨 준 후에 사라지는 과정이 반복된다. 이렇게 특정 함수호출과 관련된 프레임을 **지역 프레임**이라 
부른다.
반면에 프로그램이 실생되는 전 과정동안 살아 있는 프레임을 **전역 프레임**이라 부른다. 

아래 그림은 앞서 지역변수와 전역변수를 설명하면서 사용한 코드를 
[파이썬 튜터](http://pythontutor.com/)에서 실행한 결과를 보여준다.

#### 주의사항
* 전역변수와 함수들은 전연 프레임(global frame)에서 다룬다.
* 지역변수들은 함수가 호출되면 생성되어 지역변수들의 정보를 다루다가, 함수의 실행이 종료되면 
    모든 정보와 함께 사라진다. 
    예를 들어, 마지막 그림에서 `minutes` 값을 확인하고자 할 때 오류가 발생하는 이유가 
    `hour_to_min` 함수가 호출될 때 생성된 지역 프레임이 함수의 실행이 종료되면 사라지기 때문이라는 
    점을 눈으로 확인할 수 있다.

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var01.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var02.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var03.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var04.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var05.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var06.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var07.png" style="width:90%">
</td>
</tr>
</table>
</p>

<p>
<table cellspacing="20">
<tr>
<td>
<img src="images/local_var08.png" style="width:90%">
</td>
</tr>
</table>
</p>

## 연습문제
1. 다음 조건을 만족시키는 `right_justify`라는 함수를 정의하라.
    * 인자는 하나만 받으며, 매개변수는 `s`라 부른다.
    * 문자열 하나를 인자로 받아 실행하면 해당 문자열의 끝이 70번째 칸에 위치하도록 
        입력받은 문자열 앞에 충분한 공백이 위치하도록 출력(print)한다. 
```python
>>> right_justify('monty')
                                                                monty
```                                                                 
    힌트: 문자열 결합 및 반복, 그리고 문자열의 길이를 되돌려주는 내장함수 `len` 활용 가능.
<br><br>
1. 파이썬은 정의된 함수도 하나의 값으로 취급한다. 
    따라서 함수를 다른 함수의 인자로 사용하거나 변수에 할당되는 값으로 사용될 수 있다. 
    예를 들어, 아래에 정의된 `do_twice` 함수는 함수 `f`를 인자로 입력받으면
    그 함수를 두 번 호출하여 실행하도록 하는 함수이다. 
```python
def do_twice(f):
    f()
    f()
```
    그리고 `print_spam` 이라는 함수를 두 번 호출하도록 `do_twice`를 활용하고자 하면 
    아래와 같이 프로그램을 작성하면 된다.
    ```python
    def print_spam():
        print('spam')

    do_twice(print_spam)
    ```
    1. 이 프로그램을 직접 입력해서 실행해 보라.
    1. 아래 조건을 만족하도록 `do_twice` 함수를 수정하라. 
        * 두 개의 인자를 사용한다.
        * 첫째 매개변수는 하나의 인자를 받는 함수를 인자로 입력받는다.
        * 둘째 매개변수는 첫째 인자에 사용된 함수의 인자로 사용될 수 있는 값을 인자로 입력받는다.
        * 첫째 인자로 사용된 함수를 둘째 인자로 사용된 값을 이용하여 두 번 연속 호출한다. 
    1. 아래 조건을 만족하도록 `do_four` 함수를 수정하라. 
        * 두 개의 인자를 사용한다.
        * 첫째 매개변수는 하나의 인자를 받는 함수를 인자로 입력받는다.
        * 둘째 매개변수는 첫째 인자에 사용된 함수의 인자로 사용될 수 있는 값을 인자로 입력받는다.
        * 첫째 인자로 사용된 함수를 둘째 인자로 사용된 값을 이용하여 네 번 연속 호출한다. 
        * 앞서 정의된 `do_twice` 함수를 반드시 활용한다.
<br><br>
답: http://greenteapress.com/thinkpython2/code/do_four.py
<br><br>
1. 이번 문제는 지금까지 배운 내용을로 풀 수 있다. 
    1. 다음과 같은 격자를 그리는 함수를 작성하라.
        ```
        + - - - - + - - - - +
        |         |         |
        |         |         |
        |         |         |
        |         |         |
        + - - - - + - - - - +
        |         |         |
        |         |         |
        |         |         |
        |         |         |
        + - - - - + - - - - +
        ```
        힌트: 한 줄에 하나 이상의 값을 출력하려면 `print` 함수에 여러 인자를 사용하면 된다.
        ```python
        print('+', '-')
        ```
        `print` 함수는 기본적으로 줄바꿈을 실행한다.
        줄바꿈을 하지 않기 위해서는 에를 들어 아래와 같이 실행하면 된다.
        ```
        print('+', end=' ')
        print('-')
        ```
        위 프로그램을 실행하면 `+ -` 형식으로 출력된다. 
        <br><br>
    1. 4개의 행과 4개의 열을 갖는 비슷한 격자를 그리는 함수를 작성하라.

    답: http://greenteapress.com/thinkpython2/code/grid.py
    <br><br>
    이 연습문제는 아래 책의 Oualline에 포함된 연습문제를 응용하였다. 
    
            Practical C Programming, Third Edition, O’Reilly Media, 1997.